contracts/fee/FeeHandler.sol

```diff
@@ -3,46 +3,176 @@
+    function withdrawFees(address token, uint256 amount) external nonReentrant onlyFeeKeeper {
+        address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
+        IERC20(token).safeTransfer(receiver, amount);
     }
 
     // @dev claim fees from the specified markets
-    // @param markets the markets to claim fees from
-    // @param tokens the fee tokens to claim
-    function claimFees(
-        address[] memory markets,
-        address[] memory tokens
-    ) external nonReentrant onlyFeeKeeper {
-        if (markets.length != tokens.length) {
-            revert Errors.InvalidClaimFeesInput(markets.length, tokens.length);
+    // @param market the markets to claim fees from
+    // @param feeToken the fee tokens to claim
+    function claimFees(address market, address feeToken) external nonReentrant {
+        uint256 version;
+        uint256 buybackGmxFactor;
+        uint256 feeAmount;
+        if (market == address(0)) {
+            version = 1;
+            buybackGmxFactor = _getBuybackGmxFactor(version);
+            feeAmount = IVaultV1(vaultV1).withdrawFees(feeToken, address(this));
+            _setAvailableFeeAmounts(version, feeToken, feeAmount);
+        } else {
+            version = 2;
+            buybackGmxFactor = _getBuybackGmxFactor(version);
+            feeAmount = FeeUtils.claimFees(dataStore, eventEmitter, market, feeToken, address(this));
+            _setAvailableFeeAmounts(version, feeToken, feeAmount);
         }
+    }
 
-        address receiver = dataStore.getAddress(Keys.FEE_RECEIVER);
+    // @dev receive an amount in feeToken by depositing the batchSize amount of the buybackToken
+    // @param feeToken the token to receive with the fee amount calculated via an oracle price
+    // @param buybackToken the token to deposit in the amount of batchSize in return for fees
+    function buybackFees(address feeToken, address buybackToken) external nonReentrant {
+        uint256 batchSize = dataStore.getUint(Keys.buybackBatchAmountKey(buybackToken));
+        if (batchSize == 0) {
+            revert Errors.InvalidBuybackTokenInput(buybackToken);
+        }
+
+        uint256 minFeeTokenAmount = _getMinFeeTokenAmount(feeToken, buybackToken, batchSize);
+
+        uint256 maxPriceImpactFactor = dataStore.getUint(Keys.buybackMaxPriceImpactFactorKey(feeToken));
+        uint256 maxFeeTokenAmount = (minFeeTokenAmount * (maxPriceImpactFactor + 10000)) / 10000;
+
+        uint256 availableFeeAmount = _getAvailableFeeAmount(feeToken, buybackToken);
+
+        if (availableFeeAmount >= maxFeeTokenAmount) {
+            _buybackFees(feeToken, buybackToken, batchSize, maxFeeTokenAmount, availableFeeAmount);
+        } else if (availableFeeAmount >= minFeeTokenAmount) {
+            _buybackFees(feeToken, buybackToken, batchSize, availableFeeAmount, availableFeeAmount);
+        } else {
+            revert Errors.InsufficientFeeAmount(feeToken, buybackToken, availableFeeAmount);
```

Comment:
i don't think this is necessary, maybe let the user pass in a "minOutputAmount" and validate that the amount that would be sent to the user is >= minOutputAmount, error should be InsufficientBuybackOutputAmount to separate it from the regular fee and swap errors

Reply:
Updated to add minOutputAmount as an input param and updated the error to InsufficientBuybackOutputAmount. The error InsufficientBuybackOutputAmount doesn’t include minOutputAmount as a param because it is passed by the user to the function so they should already have that information, but it can be added as well if preferred.

Reply:
would be good to add minOutputAmount to the error for convenience, in the Errors we usually always return (..., "the actual value", "the expected value")

Reply:
Ok no problem, this has been added.